lib/jekyll/cache.rb

```diff
@@ -0,0 +1,168 @@
+# frozen_string_literal: true
+
+require "digest"
+
+module Jekyll
+  class Cache
+    # rubocop:disable Style/ClassVars
+    @@caches = {}
+    @@disk_cache_enabled = true
+
+    # Get an existing named cache, or create a new one if none exists
+    #
+    # name - name of the cache
+    #
+    # Returns nothing.
+    def initialize(name)
+      @@base_dir ||= File.expand_path(".jekyll-cache/Jekyll/Cache")
+      @cache = @@caches[name] ||= {}
+      @name = name.gsub(%r![^\w\s-]!, "-")
+    end
+
+    # Disable Marshaling cached items to disk
+    def self.disable_disk_cache!
+      @@disk_cache_enabled = false
+    end
+    # rubocop:enable Style/ClassVars
+
+    # Clear all caches
+    def self.clear
+      delete_cache_files
+      @@caches.each_value(&:clear)
+    end
+
+    # Clear this particular cache
+    def clear
+      delete_cache_files
+      @cache.clear
+    end
+
+    # Retrieve a cached item
+    # Raises if key does not exist in cache
+    #
+    # Returns cached value
+    def [](key)
+      return @cache[key] if @cache.key?(key)
+      path = path_to(hash(key))
+      if @@disk_cache_enabled && File.file?(path) && File.readable?(path)
+        @cache[key] = load(path)
+      else
+        raise
+      end
+    end
+
+    # Add an item to cache
+    #
+    # Returns nothing.
+    def []=(key, value)
+      @cache[key] = value
+      return unless @@disk_cache_enabled
+      path = path_to(hash(key))
+      dump(path, value)
+    end
+
+    # If an item already exists in the cache, retrieve it
+    # Else execute code block, and add the result to the cache, and return that
+    #   result
+    def getset(key)
+      self[key]
+    rescue StandardError
+      value = yield
+      self[key] = value
+      value
+    end
```

Comment:
can't the `rescue` block be simplified to the following?
```ruby
rescue
  self[key] = yield
end
```

Reply:
No, because we cannot assume that `[]=` returns `value`.

Reply:
I see.. Fair enough..

Reply:
Currently, our Rubocop settings forbid setting an explicit return value in `[]=`.
I did consider this though.